latency-breakdown.md

```diff
@@ -0,0 +1,799 @@
+            NonTerminal("Read index Read Wait"),
+        ),
+    ),
+    NonTerminal("Fetch A Snapshot From KV Engine"),
+)
+```
+
+```text
+tikv_storage_engine_async_request_duration_seconds{type="snapshot"} =
+    tikv_coprocessor_request_wait_seconds{type="snapshot"} =
+    tikv_raftstore_request_wait_time_duration_secs +
+    tikv_raftstore_commit_log_duration_seconds +
+    get snapshot from rocksdb duration
+```
+
+When leader lease is expired, TiKV will propose a read index command before getting a snapshot from rocksdb. `tikv_raftstore_request_wait_time_duration_secs` and `tikv_raftstore_commit_log_duration_seconds` are the duration of committing read index command.
+
+Getting snapshot from rocksdb duration is usually a fast operation, so we do not record such duration.
+
+## Async Write
+
+- Async IO Disabled
+
+```railroad+diagram
+Diagram(
+    NonTerminal("Propose Wait"),
+    NonTerminal("Process Command"),
+    Choice(
+        0,
+        Sequence(
+            NonTerminal("Wait Current Batch"),
+            NonTerminal("Write to Log Engine"),
+        ),
+        Sequence(
+            NonTerminal("RaftMsg Send Wait"),
+            NonTerminal("Commit Log Wait"),
+        ),
+    ),
+    NonTerminal("Apply Wait"),
+    NonTerminal("Apply Log"),
+)
+```
+
+- Async IO Enabled
+
+```railroad+diagram
+Diagram(
+    NonTerminal("Propose Wait"),
+    NonTerminal("Process Command"),
+    Choice(
+        0,
+        NonTerminal("Wait Until Persisted by Write Worker"),
+        Sequence(
+            NonTerminal("RaftMsg Send Wait"),
+            NonTerminal("Commit Log Wait"),
+        ),
+    ),
+    NonTerminal("Apply Wait"),
+    NonTerminal("Apply Log"),
+)
+```
+
+```text
+async write duration(async io disabled) =
+    propose +
+    async io disabled commit +
+    tikv_raftstore_apply_wait_time_duration_secs +
+    tikv_raftstore_apply_log_duration_seconds
+
+async write duration(async io enabled) =
+    propose +
+    async io enabled commit +
+    tikv_raftstore_apply_wait_time_duration_secs +
+    tikv_raftstore_apply_log_duration_seconds
+```
+
+Async write can be broken down into 3 phases:
+
+- Propose
+- Commit
+- Apply(`tikv_raftstore_apply_wait_time_duration_secs + tikv_raftstore_apply_log_duration_seconds`)
+
+```text
+propose =
+    propose wait duration +
+    propose duration
+
+propose wait duration =
+    tikv_raftstore_store_wf_batch_wait_duration_seconds
+
+propose duration =
+    tikv_raftstore_store_wf_send_to_queue_duration_seconds -
+    tikv_raftstore_store_wf_batch_wait_duration_seconds
+```
+
+The Raft process is recorded in a waterfall manner. So the propose duration is calculated from the diff of 2 metrics.
+
+```text
+async io disabled commit = max(
+    persist log locally duration,
+    replicate log duration
+)
+
+async io enabled commit = max(
+    wait by write worker duration,
+    replicate log duration
+)
+```
+
+TiKV supports Async IO Raft since v5.3.0, which changes the process of commit.
```

Comment:
Let's clarify it's not enabled by default. It's enabled when `store-io-pool-size` > 0